keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+from the kubeadm repository.
+8) Announce the move on the kubernetes-dev mailing list.
+9) Remove kubeadm and references to cmd/kubeadm from k/k.
+10) Enable the post-submit job (disable its dry-run), so that new k/k
+tags and branches can be synced in k/kubeadm.
+
+### User Stories
+
+#### Story 1
+
+As a user, I would like to build kubeadm without depending on the
+kubernetes/kubernetes repository.
+
+kubeadm no longer imports k/k packages, so this is already possible
+if one just copies cmd/kubeadm outside of k/k and creates a new Go
+module for it. It still requires some hacks around applying a version in the
+kubeadm binary. Once kubeadm moves out both the build tooling and module
+will be available to users for easy building.
+
+#### Story 2
+
+As a user, I would like to import the kubeadm public, versioned API without
+importing kubernetes/kubernetes.
+
+Currently importing the k/k root module will make your build fail.
+The best solution for users is to fork the kubeadm API packages.
+Once kubeadm moves out, importing the "kubeadm" module will make
+it possible to easily import only the API portion.
+
+#### Story 3
+
+As a user, I would like to import utility packages from kubeadm.
+
+Similar to the above user story, importing kubeadm from k/k will fail.
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository.
+
+This makes it possible to use `go get k8s.io/kubeadm` to obtain
+the kubeadm source tree.
+
+The current contents of the `k/k/cmd/kubeadm/app` directory
+should be moved in the root of k/kubeadm repository.
+
+The current `k/k/cmd/kubeadm/test` directory that contains
+integration tests should be merged with the `k/kubeadm/tests`
+directory under `k/kubeadm/tests/integration`.
+
+Existing tools in the k/kubeadm repository such as "kinder" and
+"operator" must continue existing as separate modules (sub-folders)
+inside the kubeadm module and can be obtained using
+`go get k8s.io/kubeadm/some-tool`.
+
+Existing docs and tests must co-exist in sub-folders too.
+
+Side effects:
+- All contents of k/kubeadm will be branched and tagged with kubeadm,
+but this is actually a positive change.
+- `hack/verify*` scripts for the root kubeadm folder should skip "kinder"
+and "operator". This area can see improvements in the future,
+potentially sharing scripts for all modules.
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
```

Comment:
Due to the k/release repo not being versioned along with k/k, this splitting of build tools isn't possible today.  If this KEP isn't fully implemented until the start of the 1.20 cycle in July it would be a non issue.  Likely better would be for k/release to asap get branches for release-X.YY to match k/k's branching.  Looks like for at least 1.17 and prior we'd keep the "current" build tooling, and the TBD release in which the move happens and latter would have the newer code.  
 This will increase the short term burden on SIG Release, but it is a necessary change.

Reply:
> Due to the k/release repo not being versioned along with k/k, this splitting of build tools isn't possible today. 

could you show code examples from k/release how such a version check would be disruptive?

the git-branching of k/release is much wanted. but it feels orthogonal to this proposal.
while i'm not very very familiar with the k/release code i think we can have something like this pseudo code:

```
if (currentVersionToBuild.AtLeast("v1.18") {
   // kubeadm builds from k/kubeadm
  return
}
// proceed with existing way of building
```

currently there is code in k/release that does similar.

> If this KEP isn't fully implemented until the start of the 1.20 cycle in July it would be a non issue.

is 1.20 the timeline when k/release will have branches?

on the side of kubeadm it is preferred that the work should start in 1.18 and it can be staffed by SIG Cluster Lifecycle to help with the k/release amends.

Reply:
The start of the 1.20 cycle would be the time that all supported k/k branches (1.19, 1.18, 1.17) have "the new way of building kubeadm" if that happens in 1.17.

Reply:
I overstate with "isn't possible today".  The two possibilities are that every piece of code gets a conditional around it, as you offer, or we create branches.

Reply:
me and @tpepper discussed that we can work on that during 1.19.